nova_vm/src/ecmascript/execution/environments/global_environment.rs

```diff
@@ -1,68 +1,83 @@
+    /// ### \[\[ObjectRecord\]\]
+    ///
+    /// Binding object is the global object. It contains global built-in
+    /// bindings as well as FunctionDeclaration, GeneratorDeclaration,
+    /// AsyncFunctionDeclaration, AsyncGeneratorDeclaration, and
+    /// VariableDeclaration bindings in global code for the associated realm.
+    object_record: ObjectEnvironment,
 
-    /// [[DeclarativeRecord]]
-    /// The Declaration Environment Record is inlined here.
-    declarative_record: HashMap<String, Binding>,
+    /// ### \[\[GlobalThisValue\]\]
+    ///
+    /// The value returned by this in global scope. Hosts may provide any
+    /// ECMAScript Object value.
+    pub(crate) global_this_value: Object,
 
-    /// [[VarNames]]
-    var_names: ElementsVector,
+    /// ### \[\[DeclarativeRecord\]\]
+    ///
+    /// Contains bindings for all declarations in global code for the associated
+    /// realm code except for FunctionDeclaration, GeneratorDeclaration,
+    /// AsyncFunctionDeclaration, AsyncGeneratorDeclaration, and
+    /// VariableDeclaration bindings.
+    declarative_record: DeclarativeEnvironment,
 
-    /// [[OuterEnv]]
+    /// ### \[\[VarNames\]\]
     ///
-    /// Per https://tc39.es/ecma262/#sec-the-environment-record-type-hierarchy:
-    /// > A _Global Environment Record_ is used for Script global declarations. It does not have an outer environment; its \[\[OuterEnv\]\] is null.
-    outer_env: (),
+    /// The string names bound by FunctionDeclaration, GeneratorDeclaration,
+    /// AsyncFunctionDeclaration, AsyncGeneratorDeclaration, and
+    /// VariableDeclaration declarations in global code for the associated
+    /// realm.
+    var_names: ElementsVector,
 }
 
 impl GlobalEnvironment {
-    pub(crate) fn new(global_object: Object, this_value: Object) -> Self {
-        Self {
-            object_record: global_object,
+    /// ### [9.1.2.5 NewGlobalEnvironment ( G, thisValue )](https://tc39.es/ecma262/#sec-newglobalenvironment)
+    ///
+    /// The abstract operation NewGlobalEnvironment takes arguments G (an
+    /// Object) and thisValue (an Object) and returns a Global Environment
+    /// Record.
+    pub(crate) fn new(agent: &mut Agent, global: Object, this_value: Object) -> GlobalEnvironment {
+        // 1. Let objRec be NewObjectEnvironment(G, false, null).
+        let object_record = ObjectEnvironment::new(global, false, None);
+
+        // 2. Let dclRec be NewDeclarativeEnvironment(null).
+        let declarative_record = DeclarativeEnvironment::new(None);
+
+        // 3. Let env be a new Global Environment Record.
+        GlobalEnvironment {
+            // 4. Set env.[[ObjectRecord]] to objRec.
+            object_record,
+
+            // 5. Set env.[[GlobalThisValue]] to thisValue.
             global_this_value: this_value,
-            declarative_record: Default::default(),
+
+            // 6. Set env.[[DeclarativeRecord]] to dclRec.
+            declarative_record,
```

Comment:
**nitpick:** Ah, now I remember was why I'd inlined these. It's the entirely unnecessary `OuterEnv` fields that crop up in this case :(